### PR TITLE
Remove unused dependencies, update workflow actions, simplify gitignore

### PR DESCRIPTION
**Pull request by surilindur**
Here is a small set of changes to do the following:

* Update workflow actions to newer versions to get rid of some warnings
* Add `--frozen-lockfile` flag to install step in CI to ensure the installed packages match `yarn.lock`
* Add `--ignore-scripts` flag to install step in CI, because there is a separate build step so the install one should not automatically build anything yet
* Only run tests after linting succeeds, because linting should be the cheaper task and code style should be enforced (although tslint seems deprecated in favour of ESLint)
* Remove `url-exists` and `precommit` as unused (the `pre-commit` package is the one used for the hooks, and that one still exists)
* Move the type dependencies from `dependencies` to `devDependencies`
* Simplify `.gitignore` and add the temporary test files in it

Any feedback would be welcome. Thank you.

### CODE REVIEW

**Comment by surilindur** [comment]
This should close #60 when merged, because `url-exists` is removed.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -5,40 +5,42 @@ jobs:
...
         with:
           path: '**/node_modules'
           key: ${{ runner.os }}-lint-modules-${{ hashFiles('**/yarn.lock') }}
-      - uses: actions/setup-node@v2
+      - uses: actions/setup-node@v3
         with:
-          node-version: 14.x
-      - run: yarn install
+          node-version: 20
+      - run: yarn install --frozen-lockfile --ignore-scripts
```
Comment (rubensworks) [inline .github/workflows/ci.yml:16]: Let's not do `--ignore-scripts`, it will also ignore scripts from dependencies, which we may need.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,40 +5,42 @@ jobs:
...
+      - uses: actions/cache@v3
         with:
           path: '**/node_modules'
           key: ${{ runner.os }}-lint-modules-${{ hashFiles('**/yarn.lock') }}
-      - uses: actions/setup-node@v2
+      - uses: actions/setup-node@v3
         with:
-          node-version: 14.x
-      - run: yarn install
+          node-version: 20
```
Comment (rubensworks) [inline .github/workflows/ci.yml:15]: Why not 20.x?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,40 +5,42 @@ jobs:
...
   test:
+    needs: lint
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest]
+        os:
+          - ubuntu-latest
         node-version:
-          - 14.x
```
Comment (rubensworks) [inline .github/workflows/ci.yml:25]: Why did you remove the .x?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -5,40 +5,42 @@ jobs:
...
-          - 16.x
+          - 14
+          - 16
```
Comment (rubensworks) [inline .github/workflows/ci.yml:28]: Might as well add 18 and 20 while you're at it.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -5,40 +5,42 @@ jobs:
...
-        uses: actions/checkout@v2
-      - uses: actions/cache@v2
+        uses: actions/checkout@v3
+      - uses: actions/cache@v3
         with:
           path: '**/node_modules'
           key: ${{ runner.os }}-test-modules-${{ hashFiles('**/yarn.lock') }}
       - name: Install dependencies
-        run: yarn install
+        run: yarn install --frozen-lockfile --ignore-scripts
```
Comment (rubensworks) [inline .github/workflows/ci.yml:43]: Let's not do `--ignore-scripts`.

File: package.json [hunk 6]
```diff
@@ -70,23 +70,22 @@
...
     "nock": "^13.0.0",
     "pre-commit": "^1.2.2",
-    "precommit": "^1.2.2",
     "ts-jest": "^29.0.0",
     "tslint": "^6.0.0",
     "tslint-eslint-rules": "^5.4.0",
     "typescript": "^4.6.2"
   },
   "dependencies": {
-    "@types/fs-extra": "^8.0.0",
```
Comment (rubensworks) [inline package.json:88]: Types belong in deps, not devDeps.

**Review by rubensworks — CHANGES_REQUESTED** [review]
While I welcome some of these changes, I'm not a fan of big PRs that do a bunch of stuff.

Could you either split this PR up into separate commits, or into separate PRs?

I guess the `url-exists` and `precommit` deps removal would be the most urgent.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (surilindur) [reply .github/workflows/ci.yml:28]: The tests seem to fail on 18 and 20, which is why I left them out (failing to retrieve a resource)

File: package.json (hunk 6 above)
Comment (surilindur) [reply package.json:88]: Oh, I did not know this belongs there. Thank you.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (surilindur) [reply .github/workflows/ci.yml:25]: It looks more tidy that way. The workflow still installs 14.21.3 for 14 without the x.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (surilindur) [reply .github/workflows/ci.yml:15]: It looks more tidy that way (see the comment below). For 20 it also still installs version 20.6. (and not 20.0.0 or something).

File: .github/workflows/ci.yml (hunk 1 above)
Comment (surilindur) [reply .github/workflows/ci.yml:16]: Oh okay, makes sense. I will remove the additional build step in CI, then, since it would then build the package twice: first when running yarn install (after it has installed, it builds), and a second time in the build step right after.

**Comment by surilindur** [comment]
I have removed everything else from this pull request except for the removal of `url-exists` and `precommit`, and will make another one with the unrelated changes if relevant later. Thank you for the feedback, and sorry about spending your time on this.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (rubensworks) [reply .github/workflows/ci.yml]: I would prefer to have the .x for consistency across all comunica-related workflows. (unless you're willing to change this in _all_ places)

File: .github/workflows/ci.yml (hunk 4 above)
Comment (rubensworks) [reply .github/workflows/ci.yml]: Ok, but maybe don't bump the linting task to 20 either then.

**Comment by rubensworks** [comment]
Thanks!
